File: frontend/src/engine/capabilities/insertDeleteRows.ts
/**
 * insertDeleteRows – Insert or delete rows/columns at a range.
 *
 * The range determines which rows or columns are affected (how many and where).
 * The shiftDirection controls whether to insert or delete, and the axis:
 *
 *   "down"  → insert blank rows above the range (existing rows shift down)
 *   "up"    → delete the range's rows (remaining rows shift up)
 *   "right" → insert blank columns to the left of the range (existing shift right)
 *   "left"  → delete the range's columns (remaining columns shift left)
 *
 * Examples:
 *   Insert 2 rows above row 5:  range="Sheet1!A5:A6", shiftDirection="down"
 *   Delete rows 3–5:            range="Sheet1!A3:A5", shiftDirection="up"
 *   Insert 1 column before B:   range="Sheet1!B1:B1", shiftDirection="right"
 *
 * Office.js notes:
 * - range.insert() / range.delete() are available from ExcelApi 1.1.
 * - These operations shift all data in the affected direction, preserving
 *   relative references in formulas where possible.
 */

import { CapabilityMeta, InsertDeleteRowsParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "insertDeleteRows",
  description: "Insert or delete rows/columns at a specified range",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: InsertDeleteRowsParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { range: address, shiftDirection } = params;
  const isInsert = shiftDirection === "down" || shiftDirection === "right";

  if (options.dryRun) {
    const op = isInsert ? "insert" : "delete";
    const axis = shiftDirection === "down" || shiftDirection === "up" ? "rows" : "columns";
    return { stepId: "", status: "success", message: `Would ${op} ${axis} at ${address}` };
  }

  const range = resolveRange(context, address);

  if (isInsert) {
    const direction =
      shiftDirection === "down"
        ? Excel.InsertShiftDirection.down
        : Excel.InsertShiftDirection.right;
    range.insert(direction);
  } else {
    const direction =
      shiftDirection === "up"
        ? Excel.DeleteShiftDirection.up
        : Excel.DeleteShiftDirection.left;
    range.delete(direction);
  }

  await context.sync();

  const opLabel = isInsert ? "Inserted" : "Deleted";
  const axisLabel = shiftDirection === "down" || shiftDirection === "up" ? "rows" : "columns";
  return {
    stepId: "",
    status: "success",
    message: `${opLabel} ${axisLabel} at ${address}`,
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/insert_delete_rows.py
"""insertDeleteRows — insert or delete whole rows / columns."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    op = (params.get("operation") or "").lower()
    address = params.get("range") or params.get("target")
    axis = (params.get("axis") or "rows").lower()
    if op not in ("insert", "delete") or not address:
        return {"status": "error", "message": "insertDeleteRows requires 'operation' (insert|delete) and 'range'."}

    rng = resolve_range(ctx.workbook_handle, address)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would {op} {axis} at {rng.address}."}

    whole = rng.rows if axis.startswith("row") else rng.columns
    try:
        if op == "insert":
            whole.api.Insert()
        else:
            whole.api.Delete()
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"{op} failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"{op.capitalize()}ed {axis} at {rng.address}.",
        "outputs": {"range": rng.address},
    }


registry.register("insertDeleteRows", handler, mutates=True)
